# Analysis goal 
- The goal for this project is to analyze Google Play's data to help understand what kinds of apps are likely to attract more users.
- I'll focus on free apps for this analysis.

## About the Data
As of September 2019, there were [approximately 2.8 million Android apps](https://www.statista.com/statistics/266210/number-of-available-applications-in-the-google-play-store/) on Google Play.

Collecting data for these many apps is not an easy task. So I decided to look for a data set that could help me. After some search I found two promising data sets:
- [A data set](https://www.kaggle.com/lava18/google-play-store-apps) with 10k apps collected on february 2019
- And [a data set](https://www.kaggle.com/lava18/google-play-store-apps) with 267k apps collected on april 2019


After some thought I decided to use the last one, because it has more data and was collected more recently.


## Opening the Data

In [2]:
from csv import reader

### The Google Play data set ###
opened_file = open('..\\data\\raw\\Google-Playstore-Full.csv')
read_file = reader(opened_file)
google_play = list(read_file)
google_play_header = google_play[0]
google_play_data = google_play[1:]

## Exploring the Data
To make it easier to explore the data set I created 3 functions that I will reuse throughout the project.

In [3]:
def print_header(header):
    print(header)
    print('\n')

In [4]:
def print_data(data, start = 0, end = 5):
    data_slice = data[start:end]
    for row in data_slice:
        print(row)
        print('\n')

In [5]:
def print_data_info(data):
    print('Number of rows:', len(data))
    print('Number of columns:', len(data[0]))

In [6]:
def print_data_overview():
    print_header(google_play_header)
    print_data(google_play_data)
    print_data_info(google_play_data)   

In [7]:
print_data_overview()

['App Name', 'Category', 'Rating', 'Reviews', 'Installs', 'Size', 'Price', 'Content Rating', 'Last Updated', 'Minimum Version', 'Latest Version', '', '', '', '']


['DoorDash - Food Delivery', 'FOOD_AND_DRINK', '4.548561573', '305034', '5,000,000+', 'Varies with device', '0', 'Everyone', 'March 29, 2019', 'Varies with device', 'Varies with device', '', '', '', '']


['TripAdvisor Hotels Flights Restaurants Attractions', 'TRAVEL_AND_LOCAL', '4.400671482', '1207922', '100,000,000+', 'Varies with device', '0', 'Everyone', 'March 29, 2019', 'Varies with device', 'Varies with device', '', '', '', '']


['Peapod', 'SHOPPING', '3.656329393', '1967', '100,000+', '1.4M', '0', 'Everyone', 'September 20, 2018', '5.0 and up', '2.2.0', '', '', '', '']


['foodpanda - Local Food Delivery', 'FOOD_AND_DRINK', '4.107232571', '389154', '10,000,000+', '16M', '0', 'Everyone', 'March 22, 2019', '4.2 and up', '4.18.2', '', '', '', '']


['My CookBook Pro (Ad Free)', 'FOOD_AND_DRINK', '4.647752285', '2291', 

## Data Wrangling

After a quick glance we can get some useful information about this data, like the columns that can be important ('Category', 'Rating', 'Reviews', 'Installs', 'Price', 'Content Rating').

If we pay a little more attention, we can see that the header and rows are missing the last 4 values. Since the header is mmissing too any information in these fields are meaningless. So we can start fixing that.

### Drop 4 last missing values

In [8]:
number_of_items_to_delete = 4

del google_play_header[-number_of_items_to_delete:]

for row in google_play_data:
    del row[-number_of_items_to_delete:]

In [9]:
print_data_overview()

['App Name', 'Category', 'Rating', 'Reviews', 'Installs', 'Size', 'Price', 'Content Rating', 'Last Updated', 'Minimum Version', 'Latest Version']


['DoorDash - Food Delivery', 'FOOD_AND_DRINK', '4.548561573', '305034', '5,000,000+', 'Varies with device', '0', 'Everyone', 'March 29, 2019', 'Varies with device', 'Varies with device']


['TripAdvisor Hotels Flights Restaurants Attractions', 'TRAVEL_AND_LOCAL', '4.400671482', '1207922', '100,000,000+', 'Varies with device', '0', 'Everyone', 'March 29, 2019', 'Varies with device', 'Varies with device']


['Peapod', 'SHOPPING', '3.656329393', '1967', '100,000+', '1.4M', '0', 'Everyone', 'September 20, 2018', '5.0 and up', '2.2.0']


['foodpanda - Local Food Delivery', 'FOOD_AND_DRINK', '4.107232571', '389154', '10,000,000+', '16M', '0', 'Everyone', 'March 22, 2019', '4.2 and up', '4.18.2']


['My CookBook Pro (Ad Free)', 'FOOD_AND_DRINK', '4.647752285', '2291', '10,000+', 'Varies with device', '$5.99', 'Everyone', 'April 1, 2019', 'Varies w

### Find duplicates

So here is where I hit a wall using pure python, to check for duplicates I'd have to do something like:
```python
duplicate_apps = []
unique_apps = []

for app in google_play_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
```
This would work just fine for a small data set. But in this data set it just takes too long, so for now I'll start using pandas

In [10]:
import pandas as pd

In [11]:
df = pd.read_csv('..\\data\\raw\\Google-Playstore-Full.csv', low_memory=False)

In [12]:
df.head()

App Name          Category  \
0                           DoorDash - Food Delivery    FOOD_AND_DRINK   
1  TripAdvisor Hotels Flights Restaurants Attract...  TRAVEL_AND_LOCAL   
2                                             Peapod          SHOPPING   
3                    foodpanda - Local Food Delivery    FOOD_AND_DRINK   
4                          My CookBook Pro (Ad Free)    FOOD_AND_DRINK   

        Rating  Reviews      Installs                Size  Price  \
0  4.548561573   305034    5,000,000+  Varies with device      0   
1  4.400671482  1207922  100,000,000+  Varies with device      0   
2  3.656329393     1967      100,000+                1.4M      0   
3  4.107232571   389154   10,000,000+                 16M      0   
4  4.647752285     2291       10,000+  Varies with device  $5.99   

  Content Rating        Last Updated     Minimum Version      Latest Version  \
0       Everyone      March 29, 2019  Varies with device  Varies with device   
1       Everyone      March 29, 2019  Varies with device  Varies with device   
2       Everyone  September 20, 2018          5.0 and up               2.2.0   
3       Everyone      March 22, 2019          4.2 and up              4.18.2   
4       Everyone       April 1, 2019  Varies with device  Varies with device   

  Unnamed: 11 Unnamed: 12 Unnamed: 13  Unnamed: 14  
0         NaN         NaN         NaN          NaN  
1         NaN         NaN         NaN          NaN  
2         NaN         NaN         NaN          NaN  
3         NaN         NaN         NaN          NaN  
4         NaN         NaN         NaN          NaN

In [13]:
df.drop(columns=['Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14'], inplace=True)

### Filtering only free apps

For this we'll focus on the price column, looking closer we can see a price like this '$5.99', which indicates that the entire column is composed of strings. Knowing this we can create a function to transform the price column in number.


In [14]:
def price_to_number(price):
    price = price.replace('$', '')
    return float(price)

Let's apply this function to the price column

In [15]:
df['Price'].apply(price_to_number)

ValueError: could not convert string to float: '2.4M'

Let's check what's going on

In [ ]:
df[df['Price'].str.contains(pat = 'M')]

Some kind of shift happened to the data on these rows. As consequence of this error, we'll delete these rows.

In [ ]:
#Get column indexes
indexes = df[df['Price'].str.contains(pat = 'M')].index

# Delete these row indexes from dataFrame
df.drop(indexes , inplace=True)

Let's try to apply our price_to_number function again

In [ ]:
df['Price'] = df['Price'].apply(price_to_number)

After trying to apply our function again, we just realized that we can do a better job by just getting all rows where price is '0', this will also fix problematic rows at the same time.

In [20]:
df_free = df[df['Price'] == '0']

### Removing Duplicate Entries
Now that we have only free apps we can go back to remove duplicate entries.
First we have to think about why the duplication is happening, let's take a look at some duplicate entries.

In [32]:
df_free.sort_values('App Name', ascending=True)[df_free.duplicated(subset=['App Name'], keep=False)].head(30)

App Name             Category  \
28513                    .Net Interview Questions            EDUCATION   
195447                   .Net Interview Questions            EDUCATION   
67744                              10,000 Chistes        ENTERTAINMENT   
111316                             10,000 Chistes            LIFESTYLE   
9293                         1000 English Stories            EDUCATION   
108073                       1000 English Stories            EDUCATION   
237847                 1000 Powerful Affirmations            LIFESTYLE   
230351                 1000 Powerful Affirmations  BOOKS_AND_REFERENCE   
175808                         1000 Praises Tamil            LIFESTYLE   
227862                         1000 Praises Tamil  BOOKS_AND_REFERENCE   
165690                        1000+ Hindi Stories            EDUCATION   
14242                         1000+ Hindi Stories        ENTERTAINMENT   
266254                          1008 Amman Pottri      MUSIC_AND_AUDIO   
168265                          1008 Amman Pottri        ENTERTAINMENT   
179588                          1008 Amman Pottri            LIFESTYLE   
25768                      101 Okey - ?nternetsiz           GAME_BOARD   
17366                      101 Okey - ?nternetsiz           GAME_BOARD   
131629                  10th Maths NCERT Solution  BOOKS_AND_REFERENCE   
96589                   10th Maths NCERT Solution  BOOKS_AND_REFERENCE   
12253                   10th Maths NCERT Solution  BOOKS_AND_REFERENCE   
82497          10th class maths solution in hindi            EDUCATION   
215207         10th class maths solution in hindi            EDUCATION   
133633                  12th Maths NCERT Solution  BOOKS_AND_REFERENCE   
12254                   12th Maths NCERT Solution  BOOKS_AND_REFERENCE   
16177   12th class maths solution in hindi Part-1            EDUCATION   
44987   12th class maths solution in hindi Part-1            EDUCATION   
263591                                        13!          GAME_PUZZLE   
99213                                         13!          GAME_PUZZLE   
91961                                      2 Cars          GAME_RACING   
86290                                      2 Cars       GAME_ADVENTURE   

             Rating Reviews    Installs  Size Price Content Rating  \
28513   4.599999905      35      5,000+  5.3M     0       Everyone   
195447  4.537036896      54     10,000+   20M     0       Everyone   
67744   4.434963703     961    100,000+  4.2M     0       Everyone   
111316  4.361298561    6222    500,000+  4.4M     0       Everyone   
9293     4.59296608    7137  1,000,000+  4.4M     0       Everyone   
108073            5       2      1,000+  4.0M     0       Everyone   
237847  4.252830029     265    100,000+  3.2M     0   Everyone 10+   
230351         4.25       4        100+  4.9M     0           Teen   
175808  4.810126781      79      5,000+   63M     0       Everyone   
227862  4.719745159     157     10,000+  1.9M     0       Everyone   
165690  4.322033882      59     10,000+  3.9M     0       Everyone   
14242   4.514175415   12698    500,000+   12M     0       Everyone   
266254  4.560723305     387    100,000+   19M     0       Everyone   
168265  4.312063694    1003    100,000+   49M     0       Everyone   
179588  4.296360493     577     50,000+   52M     0       Everyone   
25768   4.543731213   68315  1,000,000+   23M     0           Teen   
17366   4.400608063   10529  1,000,000+   30M     0       Everyone   
131629            5       5         10+   30M     0       Everyone   
96589   4.549529552    8611    500,000+   26M     0       Everyone   
12253    4.53656292    4335    100,000+   29M     0       Everyone   
82497   4.518574715    2638    500,000+   41M     0       Everyone   
215207  4.421052456      38     10,000+   28M     0       Everyone   
133633  4.483117104     770     50,000+   37M     0       Everyone   
12254   4.510233402    3469    100,000+   28M     0       Everyone

As we can the first apps are missing names and they were just replaced with '#NAME?'. Let's remove them.

In [30]:
#Get column indexes
indexes = df_free[df_free['App Name'] == '#NAME?'].index

# Delete these row indexes from dataFrame
df_free.drop(indexes , inplace=True)

Now to avoid removing duplicates randomly, we can remove them based on some information like rating, number of reviews, number of installs, last updated or last version. But number of reviews seems the best, the highest number of reviews among the duplicates would give us at the same time the most reliable rating, so we'll keep the highest number of reviews.

In [33]:
#First we'll sort the data frame based on the Reviews column
df_free.sort_values('Reviews', ascending=False, inplace=True)

In [34]:
#After we'll drop the duplicated names leaving only the highest number of reviews app
df_free.drop_duplicates(subset='App Name', keep='first', inplace=True)